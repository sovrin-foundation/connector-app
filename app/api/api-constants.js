// @flow

export const TOKEN_EXPIRED_CODE = 'VES-101'
export const SERVER_ERROR_CODE = 'OCS-000'
export const SERVER_API_CALL_ERROR = 'OCS-001'
export const PENDING_CONNECTION_REQUEST_CODE = 'CRS-103'
export const PUSH_NOTIFICATION_SENT_CODE = 'push-notification-sent'

export const QR_CODE_REQUEST_ID = 'id'
export const QR_CODE_SENDER_DETAIL = 's'
export const QR_CODE_SENDER_DID = 'd'
export const QR_CODE_SENDER_NAME = 'n'
export const QR_CODE_LOGO_URL = 'l'
export const QR_CODE_SENDER_VERIFICATION_KEY = 'v'
export const QR_CODE_SENDER_KEY_DELEGATION = 'dp'
export const QR_CODE_DELEGATION_DID = 'd'
export const QR_CODE_DELEGATION_KEY = 'k'
export const QR_CODE_DELEGATION_SIGNATURE = 's'
export const QR_CODE_SENDER_AGENCY = 'sa'
export const QR_CODE_SENDER_AGENCY_DID = 'd'
export const QR_CODE_SENDER_AGENCY_KEY = 'v'
export const QR_CODE_SENDER_AGENCY_ENDPOINT = 'e'
export const QR_CODE_TARGET_NAME = 't'
export const GENESIS_FILE_NAME = 'pool_transactions_genesis'

export const MESSAGE_TYPE = {
  AUTH: 'auth-req',
  CLAIM_OFFER: 'credOffer',
  CLAIM_REQUEST: 'credReq',
  PROOF_REQUEST: 'proofReq',
  PROOF: 'proof',
  CLAIM: 'cred',
}

export const PAYLOAD_TYPE = {
  GET_MESSAGE: 'GET_MSGS',
  UPDATE_PUSH_COM_METHOD: 'UPDATE_PUSH_COM_METHOD',
  SEND_MESSAGE: 'SEND_MSG',
}

export const STATUS_CODE = {
  ACCEPT: 'MS-104',
}

export const ERROR_ALREADY_EXIST = {
  code: 'OCS',
  message: 'You already have a relationship with ',
}

export const ERROR_ALREADY_EXIST_TITLE = 'Relationship already established'

export const ERROR_INVITATION_RESPONSE_PARSE_CODE = 'RN-100'

export const ERROR_INVITATION_RESPONSE_PARSE =
  'Error parsing error message in invitation response'

export const ERROR_PENDING_INVITATION_RESPONSE_PARSE_CODE = 'RN-101'

export const ERROR_PENDING_INVITATION_RESPONSE_PARSE =
  'Error parsing error message Pending Invitation SMS'

export const ERROR_INVITATION_RESPONSE_FAILED =
  'Failed to establish connection. Please try again later.'
