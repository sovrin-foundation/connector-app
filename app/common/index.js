// @flow
export * from './button-constants'
export * from './message-constants'
export * from './route-constants'
export * from './secure-storage-constants'
export function noop() {}
