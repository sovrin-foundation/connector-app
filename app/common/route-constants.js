// @flow
export const splashScreenRoute = 'SplashScreenHide'
export const homeRoute = 'Home'
export const authenticationRoute = 'Authentication'
export const connectionRoute = 'Connection'
export const settingsRoute = 'Settings'
export const settingsScreen = 'SettingsStack'
export const expiredTokenRoute = 'ExpiredToken'
export const qrCodeScannerRoute = 'QRCodeScanner'
export const lockPinSetupRoute = 'PinSetup'
export const lockTouchIdSetupRoute = 'TouchIdSetup'
export const lockPinSetupHomeRoute = 'PinSetupHome'
export const lockSelectionRoute = 'LockSelection'
export const lockEnterPinRoute = 'LockEnterPin'
export const lockSetupSuccessRoute = 'LockSetupSuccess'
export const lockEnterFingerprintRoute = 'LockEnterFingerprint'
export const claimOfferRoute = 'ClaimOffer'
export const homeTabRoute = 'TabHome'
export const connectionsTabRoute = 'TabHome'
export const credentialsTabRoute = 'TabCredentials'
export const discoverTabRoute = 'TabDiscover'
export const menuTabRoute = 'TabMenu'
export const settingsTabRoute = 'TabSettings'
export const qrCodeScannerTabRoute = 'TabQRCodeScanner'
export const proofRequestRoute = 'ProofRequest'
export const invitationRoute = 'InvitationRoute'
export const connectionHistoryRoute = 'ConnectionHistoryRoute'
export const connectionHistoryDetailsRoute = 'ConnectionHistoryDetailsRoute'
export const switchEnvironmentRoute = 'SwitchEnvironment'
export const lockAuthorizationRoute = 'LockAuthorizationRoute'
export const lockAuthorizationHomeRoute = 'LockAuthorizationHomeRoute'
export const waitForInvitationRoute = 'WaitForInvitationRoute'
export const connectionHistoryPendingRoute = 'ConnectionHistoryPendingRoute'
export const walletRoute = 'TokenWallet'
export const walletTabSendDetailsRoute = 'WalletTabSendDetails'
export const receiveTabRoute = 'WalletReceiveTab'
export const sendTabRoute = 'WalletSendTab'
export const historyTabRoute = 'WalletHistoryTab'
export const aboutAppRoute = 'AboutApp'
export const onfidoRoute = 'Onfido'
export const privacyTNCRoute = 'PrivacyTNC'
export const eulaRoute = 'EulaScreen'
export const genRecoveryPhraseRoute = 'GenerateRecoveryPhrase'
export const verifyRecoveryPhraseRoute = 'VerifyRecoveryPhrase'
export const exportBackupFileRoute = 'ExportBackupFile'
export const backupCompleteRoute = 'BackupComplete'
export const restoreRoute = 'RestoreStartScreen'
export const restoreWaitRoute = 'RestoreWaitScreen'
export const restorePassphraseRoute = 'RestorePassphrase'
export const backupErrorRoute = 'BackupErrorScreen'
