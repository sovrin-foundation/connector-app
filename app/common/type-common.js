// @flow

import type { Element } from 'react'
import type { NavigationScreenProp } from 'react-navigation'

export type CustomError = {
  code: string,
  message: string,
}

export const INITIAL_TEST_ACTION: 'INITIAL_TEST_ACTION' = 'INITIAL_TEST_ACTION'

export type InitialTestAction = {
  type: typeof INITIAL_TEST_ACTION,
}

export const initialTestAction = () => ({
  type: INITIAL_TEST_ACTION,
})

export type NavigationParams = {
  [string]: any,
}
export type NavigationRoute = NavigationLeafRoute
export type NavigationState = {
  index: number,
  routes: Array<NavigationRoute>,
}

export type NavigationLeafRoute = {|
  key: string,
  routeName: string,
  path?: string,
  params?: NavigationParams,
|}

export type ReactNavigation = NavigationScreenProp

export type GenericObject = {
  [string]: any,
}

export type GenericStringObject = {
  [string]: string,
}

export type MessageAnnotation = {
  name: string,
  version: string,
}

export type TopicAnnotation = {
  tid: number,
  mid: number,
}

export type ReactChildren = Element<*>

export const RESET: 'RESET' = 'RESET'
export const REMOVE_SERIALIZED_CLAIM_OFFERS_SUCCESS: 'REMOVE_SERIALIZED_CLAIM_OFFERS_SUCCESS' =
  'REMOVE_SERIALIZED_CLAIM_OFFERS_SUCCESS'

export type ResetAction = {
  type: typeof RESET,
}

export type ImageSource = {
  uri: string,
}

export type NotificationPayload = {
  forDID: string,
  uid: string,
  type: string,
  remotePairwiseDID: string,
  senderLogoUrl?: ?string,
}

export type ImmutableGenericStringObject = {
  +[string]: string,
}

export const STORE_STATUS = {
  IDLE: 'IDLE',
  IN_PROGRESS: 'IN_PROGRESS',
  ERROR: 'ERROR',
  SUCCESS: 'SUCCESS',
}

export type StoreError = { +error: ?CustomError }
export type StoreStatus = { +status: $Keys<typeof STORE_STATUS> }

export type StatusBarStyle = 'default' | 'light-content' | 'dark-content'
