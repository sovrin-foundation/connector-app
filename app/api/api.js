// @flow
import { api, options } from './api-utils'
import type {
  GetInvitationLinkApiData,
  MessageApiData,
  DownloadInvitationApiData,
  SendAuthenticationResponseApiData,
  EnvironmentDetailUrlDownloaded,
} from './type-api'
import { PAYLOAD_TYPE, MESSAGE_TYPE, STATUS_CODE } from './api-constants'

export const getInvitationLink = ({
  agencyUrl,
  smsToken,
}: GetInvitationLinkApiData) =>
  api(`${agencyUrl}/agency/url-mapper/${smsToken}`, options('GET'))

export const invitationDetailsRequest = ({ url }: DownloadInvitationApiData) =>
  api(url, options('GET'))

export const getAdditionalData = ({ agencyUrl, dataBody }: MessageApiData) =>
  api(`${agencyUrl}/agency/route`, options('POST', dataBody))

export const sendAuthenticationRequest = ({
  data: { identifier, dataBody },
  config: { agencyUrl },
}: SendAuthenticationResponseApiData) =>
  api(`${agencyUrl}/agent/${identifier}/auth`, options('PUT', dataBody), true)

export const downloadEnvironmentDetails = (
  url: string
): Promise<EnvironmentDetailUrlDownloaded> => api(url, options())
