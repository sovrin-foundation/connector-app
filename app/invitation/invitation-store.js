// @flow
import { Platform } from 'react-native'
import { put, takeLatest, call, all, select } from 'redux-saga/effects'
import {
  INVITATION_RECEIVED,
  INVITATION_RESPONSE_SEND,
  INVITATION_RESPONSE_SUCCESS,
  INVITATION_RESPONSE_FAIL,
  INVITATION_REJECTED,
  ERROR_INVITATION_VCX_INIT,
  ERROR_INVITATION_CONNECT,
} from './type-invitation'
import { ResponseType } from '../components/request/type-request'
import {
  ERROR_ALREADY_EXIST,
  ERROR_INVITATION_RESPONSE_PARSE_CODE,
  ERROR_INVITATION_RESPONSE_PARSE,
  SERVER_ERROR_CODE,
} from '../api/api-constants'
import {
  getAgencyUrl,
  getAgencyDID,
  getAgencyVerificationKey,
  getPushToken,
  getInvitationPayload,
  isDuplicateConnection,
  getUserOneTimeInfo,
  getPoolConfig,
} from '../store/store-selector'
import { saveNewConnection } from '../store/connections-store'
import {
  createConnectionWithInvite,
  acceptInvitationVcx,
  serializeConnection,
} from '../bridge/react-native-cxs/RNCxs'
import type {
  InvitationResponseSendData,
  InvitationResponseSendAction,
  InvitationPayload,
  InvitationStore,
  InvitationAction,
  InvitationReceivedActionData,
} from './type-invitation'
import type { CustomError } from '../common/type-common'
import { captureError } from '../services/error/error-handler'
import type {
  ConnectAgencyResponse,
  RegisterAgencyResponse,
  CreateOneTimeAgentResponse,
  CreatePairwiseAgentResponse,
  AcceptInvitationResponse,
} from '../bridge/react-native-cxs/type-cxs'
import type { UserOneTimeInfo } from '../store/user/type-user-store'
import { connectRegisterCreateAgentDone } from '../store/user/user-store'
import { RESET } from '../common/type-common'
import { ensureVcxInitSuccess } from '../store/config-store'
import { VCX_INIT_SUCCESS } from '../store/type-config-store'
import type { MyPairwiseInfo } from '../store/type-connection-store'
import { safeSet, safeGet } from '../services/storage'

export const invitationInitialState = {}

export const invitationReceived = (data: InvitationReceivedActionData) => ({
  type: INVITATION_RECEIVED,
  data,
})

export const sendInvitationResponse = (data: InvitationResponseSendData) => ({
  type: INVITATION_RESPONSE_SEND,
  data,
})

export const invitationSuccess = (senderDID: string) => ({
  type: INVITATION_RESPONSE_SUCCESS,
  senderDID,
})

export const invitationFail = (error: CustomError, senderDID: string) => ({
  type: INVITATION_RESPONSE_FAIL,
  error,
  senderDID,
})

export const invitationRejected = (senderDID: string) => ({
  type: INVITATION_REJECTED,
  senderDID,
})

export function* sendResponse(
  action: InvitationResponseSendAction
): Generator<*, *, *> {
  const { senderDID } = action.data

  try {
    const vcxResult = yield* ensureVcxInitSuccess()
    const alreadyExist: boolean = yield select(isDuplicateConnection, senderDID)
    if (alreadyExist) {
      yield put(invitationFail(ERROR_ALREADY_EXIST, senderDID))

      return
    }
    if (vcxResult && vcxResult.fail) {
      throw new Error(vcxResult.fail.message)
    }
    const payload: InvitationPayload = yield select(
      getInvitationPayload,
      senderDID
    )
    const connectionHandle: number = yield call(
      createConnectionWithInvite,
      payload
    )
    const pairwiseInfo: MyPairwiseInfo = yield call(
      acceptInvitationVcx,
      connectionHandle
    )

    yield put(invitationSuccess(senderDID))

    // once the connection is successful, we need to save serialized connection
    // in secure storage as well, because libIndy does not handle persistence
    // once we have persisted serialized state, we can hydrate vcx
    // if we need anything from that connection
    const vcxSerializedConnection: string = yield call(
      serializeConnection,
      connectionHandle
    )
    // TODO:KS Above call will probably interfere with showing bubbles as soon as user
    // is on home screen after accepting connection
    // there are few more things that we can do, but we are not doing those here for now

    const connection = {
      newConnection: {
        identifier: pairwiseInfo.myPairwiseDid,
        logoUrl: payload.senderLogoUrl,
        myPairwiseDid: pairwiseInfo.myPairwiseDid,
        myPairwiseVerKey: pairwiseInfo.myPairwiseVerKey,
        myPairwiseAgentDid: pairwiseInfo.myPairwiseAgentDid,
        myPairwiseAgentVerKey: pairwiseInfo.myPairwiseAgentVerKey,
        myPairwisePeerVerKey: pairwiseInfo.myPairwisePeerVerKey,
        vcxSerializedConnection,
        ...payload,
      },
    }
    yield put(saveNewConnection(connection))
  } catch (e) {
    captureError(e)
    yield put(invitationFail(ERROR_INVITATION_CONNECT(e.message), senderDID))
  }
}

function* watchSendInvitationResponse(): any {
  yield takeLatest(INVITATION_RESPONSE_SEND, sendResponse)
}

export function* watchInvitation(): any {
  yield all([watchSendInvitationResponse()])
}

export default function invitationReducer(
  state: InvitationStore = invitationInitialState,
  action: InvitationAction
) {
  switch (action.type) {
    case INVITATION_RECEIVED:
      return {
        ...state,
        [action.data.payload.senderDID]: {
          ...action.data,
          status: ResponseType.none,
          isFetching: false,
          error: null,
        },
      }

    case INVITATION_RESPONSE_SEND:
      return {
        ...state,
        [action.data.senderDID]: {
          ...state[action.data.senderDID],
          isFetching: true,
          status: action.data.response,
          error: null,
        },
      }

    case INVITATION_RESPONSE_SUCCESS:
      return {
        ...state,
        [action.senderDID]: {
          ...state[action.senderDID],
          isFetching: false,
          error: null,
        },
      }

    case INVITATION_RESPONSE_FAIL:
      return {
        ...state,
        [action.senderDID]: {
          ...state[action.senderDID],
          isFetching: false,
          error: action.error,
          status: ResponseType.none,
        },
      }

    case INVITATION_REJECTED:
      return {
        ...state,
        [action.senderDID]: {
          ...state[action.senderDID],
          isFetching: false,
          error: null,
          status: ResponseType.rejected,
        },
      }

    case RESET:
      return invitationInitialState

    default:
      return state
  }
}
