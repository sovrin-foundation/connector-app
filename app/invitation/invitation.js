// @flow
import React, { PureComponent } from 'react'
import { View, StatusBar } from 'react-native'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { captureError } from '../services/error/error-handler'
import {
  Request,
  Container,
  CustomModal,
  Loader,
  CustomText,
  CustomView,
} from '../components'
import { homeRoute } from '../common'
import { OFFSET_1X } from '../common/styles'
import { ResponseType } from '../components/request/type-request'
import { sendInvitationResponse, invitationRejected } from './invitation-store'
import type { Store } from '../store/type-store'
import type { ResponseTypes } from '../components/request/type-request'
import type { InvitationProps, InvitationState } from './type-invitation'
import type { ReactNavigation } from '../common/type-common'
import { smsPendingInvitationSeen } from '../sms-pending-invitation/sms-pending-invitation-store'
import { SMSPendingInvitationStatus } from '../sms-pending-invitation/type-sms-pending-invitation'
import { NavigationActions } from 'react-navigation'
import { barStyleDark, color } from '../common/styles/constant'

export class Invitation extends PureComponent<
  InvitationProps,
  InvitationState
> {
  state = {
    loading: false,
  }

  isLoading = (state: boolean) => this.setState({ loading: state })

  navigate = () => {
    this.props.navigation.navigate(homeRoute)
  }

  onSuccess = () => {
    setTimeout(() => {
      this.navigate()
    }, 500)
  }

  onAction = (response: ResponseTypes) => {
    const { invitation } = this.props
    if (invitation) {
      const { payload } = invitation
      if (payload) {
        if (response === ResponseType.accepted) {
          this.props.sendInvitationResponse({
            response,
            senderDID: payload.senderDID,
          })
        } else if (response === ResponseType.rejected) {
          this.props.invitationRejected(payload.senderDID)
          this.navigate()
        }
      } else {
        this.onSuccess()
      }
    }
  }

  componentWillReceiveProps(nextProps: InvitationProps) {
    // If invitation itself not generated then don't check payload
    if (nextProps.invitation && this.props.invitation) {
      if (
        this.props.invitation !== undefined &&
        nextProps.invitation.payload !== this.props.invitation.payload
      ) {
        // a new invitation was received
        this.isLoading(false)
      } else {
        if (nextProps.invitation.isFetching === false) {
          if (nextProps.invitation.error) {
            // TODO:KS we got error from API response, what to do now
            if (nextProps.invitation.error != this.props.invitation.error) {
              // TODO: captureError should only accept Error object or only json
              // as of now it would fail for JSON objects, as we are using
              // captureException of Sentry which is supposed to take only Error object
              // captureError(nextProps.invitation.error, this.props.showErrorAlerts)
              this.isLoading(false)
            }
          } else {
            // api response was successful, but now we have to check
            // if user accepted or declined the request
            if (nextProps.invitation.status === ResponseType.accepted) {
              this.isLoading(false)
              this.onSuccess()
            }
          }
        } else {
          // TODO:KS show loading indicator, API request was sent
          this.isLoading(true)
        }
      }
    }
  }

  componentDidMount() {
    if (this.props.isSmsInvitationNotSeen) {
      this.props.smsPendingInvitationSeen(this.props.smsToken)
    }
  }

  render() {
    const { invitation, showErrorAlerts, navigation } = this.props

    if (invitation) {
      let { payload } = invitation
      let senderName = ''
      let title = 'Hi'
      let message = 'You have received a connection request'
      let senderLogoUrl = undefined
      if (payload) {
        senderName = payload.senderName
        title = `Hi ${payload.targetName}`
        message = `${senderName} wants to connect with you.`
        senderLogoUrl = payload.senderLogoUrl
        return (
          <Container>
            <StatusBar
              barStyle={barStyleDark}
              animated={true}
              backgroundColor={color.bg.fifth.color}
            />
            <Request
              title={title}
              message={message}
              senderLogoUrl={senderLogoUrl}
              onAction={this.onAction}
              testID={'invitation'}
              navigation={navigation}
              showErrorAlerts={showErrorAlerts}
              invitationError={invitation.error}
              senderName={senderName}
            />
            {this.state.loading && (
              <CustomModal
                testID={'invitation'}
                isVisible={this.state.loading}
                fullScreen
              >
                <CustomView center style={[{ height: '90%' }]}>
                  <CustomText
                    h5
                    center
                    tertiary
                    bg="tertiary"
                    transparentBg
                    style={[{ marginBottom: OFFSET_1X / 2 }]}
                    bold
                  >
                    Connecting...
                  </CustomText>
                  <Loader />
                </CustomView>
              </CustomModal>
            )}
          </Container>
        )
      }
    }
    return <Container />
  }
}

const mapStateToProps = (state: Store, { navigation }: ReactNavigation) => {
  const senderDID = navigation.state ? navigation.state.params.senderDID : ''
  const smsToken = navigation.state ? navigation.state.params.token : null
  const isSmsInvitationNotSeen =
    smsToken &&
    state.smsPendingInvitation[smsToken] &&
    state.smsPendingInvitation[smsToken].status !==
      SMSPendingInvitationStatus.SEEN
  return {
    invitation: state.invitation[senderDID],
    showErrorAlerts: state.config.showErrorAlerts,
    smsToken,
    isSmsInvitationNotSeen,
  }
}

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    { sendInvitationResponse, invitationRejected, smsPendingInvitationSeen },
    dispatch
  )

export default connect(mapStateToProps, mapDispatchToProps)(Invitation)
