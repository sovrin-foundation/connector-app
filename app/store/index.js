// @flow
import { createStore, combineReducers, applyMiddleware } from 'redux'
import createSagaMiddleware from 'redux-saga'
import { all } from 'redux-saga/effects'
import logger from 'redux-logger'
import user, { watchUserStore } from './user/user-store'
import pushNotification, {
  watchPushNotification,
} from '../push-notification/push-notification-store'
import connections, { watchConnection } from './connections-store'
import config, { watchConfig, watchGetMessagesSaga } from './config-store'
import authentication, {
  watchAuthentication,
} from '../authentication/authentication-store'
import deepLink from '../deep-link/deep-link-store'
import route from './route-store'
import lock, {
  watchLock,
  watchPressEventInLockSelectionScreen,
  watchEnableTouchId,
  watchDisableTouchId,
} from '../lock/lock-store'
import smsPendingInvitation, {
  watchSmsPendingInvitationSaga,
} from '../sms-pending-invitation/sms-pending-invitation-store'
import claimOffer, { watchClaimOffer } from '../claim-offer/claim-offer-store'
import proofRequest, {
  watchProofRequestAccepted,
  watchPersistProofRequests,
  watchProofRequestReceived,
} from '../proof-request/proof-request-store'
import invitation, { watchInvitation } from '../invitation/invitation-store'
import claim, { watchClaim } from '../claim/claim-store'
import proof, { watchProof } from '../proof/proof-store'
import history, {
  watchConnectionHistory,
} from '../connection-history/connection-history-store'
import historyRecorder from '../connection-history/history-middleware'
import wallet, { watchWalletStore } from '../wallet/wallet-store'
import eula, { watchEula } from '../eula/eula-store'
import restore, { watchRestore } from '../restore/restore-store'
import backup, { watchBackup } from '../backup/backup-store'
import onfido, { watchOnfido } from '../onfido/onfido-store'
import offline, { watchOffline } from '../offline/offline-store'
import { hydrate } from './hydration-store'
import {
  watchLedgerStore,
  ledgerStoreReducer as ledger,
} from './ledger/ledger-store'

const sagaMiddleware = createSagaMiddleware()

const appReducer = combineReducers({
  config,
  connections,
  deepLink,
  authentication,
  pushNotification,
  route,
  smsPendingInvitation,
  user,
  lock,
  claimOffer,
  proofRequest,
  invitation,
  claim,
  proof,
  history,
  wallet,
  eula,
  restore,
  backup,
  ledger,
  offline,
  onfido,
})

let middlewares = [historyRecorder]

if (process.env.NODE_ENV !== 'test' && __DEV__) {
  // skip logger middleware if we are running tests
  middlewares.push(logger)
}

middlewares.push(sagaMiddleware)

const store = createStore(appReducer, applyMiddleware(...middlewares))

// $FlowFixMe Don't know how to fix polymorphic type coercion
sagaMiddleware.run(function*() {
  return yield all([
    watchConnection(),
    watchConfig(),
    watchAuthentication(),
    watchLock(),
    watchSmsPendingInvitationSaga(),
    watchClaimOffer(),
    watchPushNotification(),
    watchInvitation(),
    watchClaim(),
    watchPressEventInLockSelectionScreen(),
    watchEnableTouchId(),
    watchDisableTouchId(),
    watchProof(),
    watchProofRequestAccepted(),
    watchConnectionHistory(),
    watchUserStore(),
    watchWalletStore(),
    watchBackup(),
    watchEula(),
    watchRestore(),
    hydrate(),
    watchGetMessagesSaga(),
    watchPersistProofRequests(),
    watchProofRequestReceived(),
    watchLedgerStore(),
    watchOffline(),
  ])
})

export default store
