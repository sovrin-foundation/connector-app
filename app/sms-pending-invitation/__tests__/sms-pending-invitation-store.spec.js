// @flow
import { call, select, put, take } from 'redux-saga/effects'
import {
  getAgencyUrl,
  getCurrentScreen,
  getHydrationState,
} from '../../store/store-selector'
import smsPendingInvitationReducer, {
  callSmsPendingInvitationRequest,
  convertSmsPayloadToInvitation,
  smsPendingInvitationFail,
  smsPendingInvitationReceived,
  getSmsPendingInvitation,
  smsPendingInvitationSeen,
} from '../sms-pending-invitation-store'
import {
  SMSPendingInvitationStatus,
  SAFE_TO_DOWNLOAD_SMS_INVITATION,
} from '../type-sms-pending-invitation'
import { getInvitationLink, invitationDetailsRequest } from '../../api/api'
import { PENDING_CONNECTION_REQUEST_CODE } from '../../api/api-constants'
import { initialTestAction } from '../../common/type-common'
import { invitationReceived } from '../../invitation/invitation-store'
import {
  smsDownloadedPayload as payload,
  smsToken,
  getStore,
} from '../../../__mocks__/static-data'
import { HYDRATED } from '../../store/type-config-store'
import { lockSelectionRoute } from '../../common/route-constants'

describe('SMS Connection Request store', () => {
  const initialState = {}
  const smsToken = 'gm76ku'
  const store = getStore()

  const getPendingInvitationState = state =>
    smsPendingInvitationReducer(state, getSmsPendingInvitation(smsToken))

  it('should be correct initial state', () => {
    const actualInitialState = smsPendingInvitationReducer(
      undefined,
      initialTestAction()
    )
    expect(actualInitialState).toEqual(initialState)
  })

  it('should update store when pending invitation is requested', () => {
    const nextState = getPendingInvitationState(initialState)

    expect(nextState).toMatchSnapshot()
  })

  it('should update store when invitation received', () => {
    const invitationRequestedState = getPendingInvitationState(initialState)
    const nextState = smsPendingInvitationReducer(
      invitationRequestedState,
      smsPendingInvitationReceived(smsToken, payload)
    )

    expect(nextState).toMatchSnapshot()
  })

  it('should update store when invitation receive fail', () => {
    const invitationRequestedState = getPendingInvitationState(initialState)
    const nextState = smsPendingInvitationReducer(
      invitationRequestedState,
      smsPendingInvitationFail(smsToken, {
        code: 'TEST-FAIL',
        message: 'Test fail message',
      })
    )

    expect(nextState).toMatchSnapshot()
  })

  it('should update store when invitation is seen', () => {
    const invitationRequestedState = getPendingInvitationState(initialState)
    const afterReceived = smsPendingInvitationReducer(
      invitationRequestedState,
      smsPendingInvitationReceived(smsToken, payload)
    )
    const nextState = smsPendingInvitationReducer(
      afterReceived,
      smsPendingInvitationSeen(smsToken)
    )

    expect(nextState).toMatchSnapshot()
  })

  it('sms invitation download workflow should work fine if api returns success', () => {
    const gen = callSmsPendingInvitationRequest(
      getSmsPendingInvitation(smsToken)
    )

    expect(gen.next().value).toEqual(select(getCurrentScreen))
    // we are explicitly adding test dependency on getCurrentScreen
    // so that if someone changes this selector or store state/structure
    // they should know that this part of app breaks if we change this selector
    expect(getCurrentScreen(store.getState())).toMatchSnapshot()

    expect(gen.next(lockSelectionRoute).value).toEqual(
      take(SAFE_TO_DOWNLOAD_SMS_INVITATION)
    )

    expect(gen.next().value).toEqual(select(getHydrationState))
    // again add explicit test dependency on selector, so that this test breaks
    // if any change is done for selector or store/structure
    expect(getHydrationState(store.getState())).toMatchSnapshot()

    expect(gen.next(false).value).toEqual(take(HYDRATED))

    expect(gen.next().value).toEqual(select(getAgencyUrl))
    const agencyUrl = 'http://test-agency.com'

    expect(gen.next(agencyUrl).value).toEqual(
      call(getInvitationLink, {
        agencyUrl,
        smsToken,
      })
    )

    const invitationData = { url: 'http://enterprise-agency.com' }

    expect(gen.next(invitationData).value).toEqual(
      call(invitationDetailsRequest, {
        url: invitationData.url,
      })
    )
    expect(gen.next(payload).value).toEqual(
      put(
        invitationReceived({
          payload: convertSmsPayloadToInvitation(payload),
        })
      )
    )

    expect(gen.next().value).toEqual(
      put(smsPendingInvitationReceived(smsToken, payload))
    )

    expect(gen.next().done).toBe(true)
  })

  //TODO : need to add test cases for multiple test cases whhen code is refactored.

  it('sms invitation download error should raise fail action', () => {
    const error = {
      code: 'OCS',
      message: 'sms connection pending request api error',
    }

    const gen = callSmsPendingInvitationRequest(
      getSmsPendingInvitation(smsToken)
    )

    // Need to find better way to share common tests for sagas
    // and common way forward

    expect(gen.next().value).toEqual(select(getCurrentScreen))
    // we are explicitly adding test dependency on getCurrentScreen
    // so that if someone changes this selector or store state/structure
    // they should know that this part of app breaks if we change this selector
    expect(getCurrentScreen(store.getState())).toMatchSnapshot()

    expect(gen.next(lockSelectionRoute).value).toEqual(
      take(SAFE_TO_DOWNLOAD_SMS_INVITATION)
    )

    expect(gen.next().value).toEqual(select(getHydrationState))
    // again add explicit test dependency on selector, so that this test breaks
    // if any change is done for selector or store/structure
    expect(getHydrationState(store.getState())).toMatchSnapshot()

    expect(gen.next(false).value).toEqual(take(HYDRATED))

    expect(gen.next().value).toEqual(select(getAgencyUrl))
    const agencyUrl = 'http://test-agency.com'

    expect(gen.next(agencyUrl).value).toEqual(
      call(getInvitationLink, {
        agencyUrl,
        smsToken,
      })
    )

    expect(gen.throw(new Error(JSON.stringify(error))).value).toEqual(
      put(smsPendingInvitationFail(smsToken, error))
    )

    expect(gen.next().done).toBe(true)
  })

  it('should reset store, if RESET action is raised', () => {
    const afterInvitationReceivedState = getPendingInvitationState(initialState)
    expect(
      smsPendingInvitationReducer(afterInvitationReceivedState, {
        type: 'RESET',
      })
    ).toMatchSnapshot()
  })
})
