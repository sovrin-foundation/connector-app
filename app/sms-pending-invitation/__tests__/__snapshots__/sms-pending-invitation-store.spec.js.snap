// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SMS Connection Request store should reset store, if RESET action is raised 1`] = `Object {}`;

exports[`SMS Connection Request store should update store when invitation is seen 1`] = `
Object {
  "gm76ku": Object {
    "error": null,
    "isFetching": false,
    "payload": Object {
      "connReqId": "requestId1",
      "senderAgencyDetail": Object {
        "DID": "senderDID1",
        "endpoint": "endpoint",
        "verKey": "senderVerificationKey1",
      },
      "senderDetail": Object {
        "DID": "senderDID1",
        "agentKeyDlgProof": Object {
          "agentDID": "senderAgentDID1",
          "agentDelegatedKey": "agentDelegatedKey",
          "signature": "signature",
        },
        "logoUrl": "http://testissuer.com/logoUrl.png",
        "name": "sender1",
        "verKey": "senderVerificationKey1",
      },
      "statusCode": "MS-102",
      "statusMsg": "message sent",
      "targetName": "target name",
    },
    "status": "SEEN",
  },
}
`;

exports[`SMS Connection Request store should update store when invitation receive fail 1`] = `
Object {
  "gm76ku": Object {
    "error": Object {
      "code": "TEST-FAIL",
      "message": "Test fail message",
    },
    "isFetching": false,
    "payload": null,
    "status": "FETCH_FAILED",
  },
}
`;

exports[`SMS Connection Request store should update store when invitation received 1`] = `
Object {
  "gm76ku": Object {
    "error": null,
    "isFetching": false,
    "payload": Object {
      "connReqId": "requestId1",
      "senderAgencyDetail": Object {
        "DID": "senderDID1",
        "endpoint": "endpoint",
        "verKey": "senderVerificationKey1",
      },
      "senderDetail": Object {
        "DID": "senderDID1",
        "agentKeyDlgProof": Object {
          "agentDID": "senderAgentDID1",
          "agentDelegatedKey": "agentDelegatedKey",
          "signature": "signature",
        },
        "logoUrl": "http://testissuer.com/logoUrl.png",
        "name": "sender1",
        "verKey": "senderVerificationKey1",
      },
      "statusCode": "MS-102",
      "statusMsg": "message sent",
      "targetName": "target name",
    },
    "status": "RECEIVED",
  },
}
`;

exports[`SMS Connection Request store should update store when pending invitation is requested 1`] = `
Object {
  "gm76ku": Object {
    "error": null,
    "isFetching": true,
    "payload": null,
    "status": "NONE",
  },
}
`;

exports[`SMS Connection Request store sms invitation download error should raise fail action 1`] = `"TabQRCodeScanner"`;

exports[`SMS Connection Request store sms invitation download error should raise fail action 2`] = `false`;

exports[`SMS Connection Request store sms invitation download workflow should work fine if api returns success 1`] = `"TabQRCodeScanner"`;

exports[`SMS Connection Request store sms invitation download workflow should work fine if api returns success 2`] = `false`;
